httpcore/_async/http11.py

```diff
@@ -319,7 +320,8 @@ async def __aiter__(self) -> AsyncIterator[bytes]:
             # If we get an exception while streaming the response,
             # we want to close the response (and possibly the connection)
             # before raising that exception.
-            await self.aclose()
+            with AsyncShieldCancellation():
+                await self.aclose()
```

Comment:
It is risky to shield this operation without a timeout because it can hang indefinitely, causing our program to hang even if it is cancelled.

Reply:
> because it can hang indefinitely

Is that true for the network closes?

Reply:
If it's a TLS connection then it could wait for shutdown

Reply:
In trio - yes.
Before opening #719, I asked the author of trio.

Reply:
I think what you want to do is handle this aclose() in a background TaskGroup

Reply:
From docs

> For example, suppose we call [aclose()](https://trio.readthedocs.io/en/stable/reference-io.html?highlight=Async%20resource#trio.abc.AsyncResource.aclose) on a TLS-encrypted connection. This requires sending a “goodbye” message; but if the peer has become non-responsive, then our attempt to send this message might block forever, and eventually time out and be cancelled. In this case the [aclose()](https://trio.readthedocs.io/en/stable/reference-io.html?highlight=Async%20resource#trio.abc.AsyncResource.aclose) method on [SSLStream](https://trio.readthedocs.io/en/stable/reference-io.html?highlight=Async%20resource#trio.SSLStream) will immediately close the underlying transport stream using [trio.aclose_forcefully()](https://trio.readthedocs.io/en/stable/reference-io.html?highlight=Async%20resource#trio.aclose_forcefully) before raising [Cancelled](https://trio.readthedocs.io/en/stable/reference-core.html#trio.Cancelled).

Reply:
Even if we call `aclose` from a cancelled environment, the resource will be closed anyway. The main question is which timeout is appropriate to wait for a graceful tear down.

Reply:
> This requires sending a “goodbye” message

Ah, we're on safe ground here...

https://github.com/python-trio/trio/issues/412#issuecomment-359888397

> for an http client, calling aclose always closes the connection immediately (because you'll be using SSLStream in https_compatible=True mode), which is both fastest and nicest to the server.

Reply:
I experimented with the trio `aclose` and discovered something strange. It always sends `RST` packets when closing `TLS` connections, regardless of whether `https_compatible` is true or false, meaning that `aclose` would never block.

Reply:
@tomchristie Take a look at [this code](https://gist.github.com/karosis88/fe4c889df9c6e9a8bfcde7dd615a080c)

Reply:
Okay, that's really helpful...

We can start to turn that into test cases.

Here's a rough sketch of something that we want to pass:

```python
def main():
    context = ssl.create_default_context()
    context.load_verify_locations('client.pem')
    timeout = {'timeout': {'connect': 5, 'read': 5, 'write': 5, 'pool': 5}}
    print('connection pool opened')
    with httpcore.ConnectionPool(ssl_context=context) as http:
        print('sending request')
        try:
            r = http.request('GET', 'https://127.0.0.1:5051/', extensions=timeout)
        except httpcore.TimeoutException:
            print('timed out')
        else:
            print('completed')
        assert http.connections == []
    print('connection pool closed')
```

Thoughts here...

* We want the default `timeout` extensions to function correctly. For `sync`, `trio`, and `asyncio` cases. With the above case (and async equivalents) they work as expected.
* We also want the above to work when using `async` cancel timeout primitives.

Reply:
> We also want the above to work when using async cancel timeout primitives.

Let's add support for `timeout` argument to our AsyncShieldCancellation, which will act as a `deadline` in trio.CancelScope and anyio.CancelScope.
To convert integer timeout to clock time, we can use the `anyio.current_time` and `trio.current_time functions`.

Exmaple:

``` python
with AsyncShieldCancellation(timeout=5):
    ...
```

Reply:
> @tomchristie Take a look at [this code](https://gist.github.com/karosis88/fe4c889df9c6e9a8bfcde7dd615a080c)

I thought this gist was convincing enough to believe it could hang.

Reply:
Are you requesting a failing test case? Should we demonstrate the test case first, or write some pseudo code that could be our solution?

Reply:
> I thought this gist was convincing enough to believe it could hang.

The following shows `httpcore` working as desired...

```python
import anyio
import ssl
import socket
import threading
import time
import httpcore

PORT = 5051

def serve():
    sock = socket.socket()
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    sock.bind(('127.0.0.1', PORT))
    sock.listen()
    context = ssl.SSLContext(protocol=ssl.PROTOCOL_TLS_SERVER)
    context.keylog_filename = 'keylogs'
    context.load_cert_chain(certfile='server.pem', keyfile='server.key')
    tls = context.wrap_socket(sock, server_side=True) 
    c, _ = tls.accept()
    time.sleep(10000)

threading.Thread(target=serve).start()
time.sleep(0.1)


async def main():
    context = ssl.create_default_context()
    context.load_verify_locations('client.pem')
    print('connection pool opened')
    async with httpcore.AsyncConnectionPool(ssl_context=context) as http:
        print('sending request')
        async with anyio.move_on_after(5):
            r = await http.request('GET', 'https://127.0.0.1:5051/')
        print(http.connections)
    print('connection pool closed')

anyio.run(main)
```

On this branch the timeout works and the connection has been removed from the pool.
On master the timeout works, but the connection remains stuck in the pool.

So, it's not obvious that we've got a remaining issue to address here?

Reply:
With DEBUG logs...

```
DEBUG [2023-07-03 11:54:22] asyncio - Using selector: KqueueSelector
connection pool opened
sending request
DEBUG [2023-07-03 11:54:22] httpcore.connection - connect_tcp.started host='127.0.0.1' port=5051 local_address=None timeout=None socket_options=None
DEBUG [2023-07-03 11:54:22] httpcore.connection - connect_tcp.complete return_value=<httpcore._backends.anyio.AnyIOStream object at 0x1092405b0>
DEBUG [2023-07-03 11:54:22] httpcore.connection - start_tls.started ssl_context=<ssl.SSLContext object at 0x1092006c0> server_hostname='127.0.0.1' timeout=None
DEBUG [2023-07-03 11:54:22] httpcore.connection - start_tls.complete return_value=<httpcore._backends.anyio.AnyIOStream object at 0x1091fda80>
DEBUG [2023-07-03 11:54:22] httpcore.http11 - send_request_headers.started request=<Request [b'GET']>
DEBUG [2023-07-03 11:54:22] httpcore.http11 - send_request_headers.complete
DEBUG [2023-07-03 11:54:22] httpcore.http11 - send_request_body.started request=<Request [b'GET']>
DEBUG [2023-07-03 11:54:22] httpcore.http11 - send_request_body.complete
DEBUG [2023-07-03 11:54:22] httpcore.http11 - receive_response_headers.started request=<Request [b'GET']>
DEBUG [2023-07-03 11:54:27] httpcore.http11 - receive_response_headers.failed exception=CancelledError()
DEBUG [2023-07-03 11:54:27] httpcore.http11 - response_closed.started
DEBUG [2023-07-03 11:54:27] httpcore.http11 - response_closed.complete
[]
connection pool closed
```

Reply:
It's my fault. It appears that my gist simply ignores the fact that `standard_compatible` should be set to False in order to avoid the closing handshake.

Reply:
Ah that makes sense now then. Okay, I think we're good to go on this one then.